src/main/java/com/hierynomus/smbj/session/Session.java

```diff
@@ -72,6 +72,37 @@ public Session(Connection connection, AuthenticationContext auth, SMBEventBus bu
     public void init(SMB2SessionSetup setup) {
         this.guest = setup.getSessionFlags().contains(SMB2SessionSetup.SMB2SessionFlags.SMB2_SESSION_FLAG_IS_GUEST);
         this.anonymous = setup.getSessionFlags().contains(SMB2SessionSetup.SMB2SessionFlags.SMB2_SESSION_FLAG_IS_NULL);
+        validateAndSetSigning(setup);
+        if (guest || anonymous) {
+            packetSignatory.init(null); // De-initialize the signatory if it was initialized with a bogus key during the auth
+        }
+    }
+
+    /**
+     * [MS-SMB2] 3.2.5.3.1 Handling a New Authentication
+     * @param setup
+     */
+    private void validateAndSetSigning(SMB2SessionSetup setup) {
+        boolean requireMessageSigning = connection.getConfig().isSigningRequired();
+        boolean connectionSigningRequired = connection.getConnectionInfo().isServerRequiresSigning();
+
+        if (requireMessageSigning || connectionSigningRequired) {
+            signingRequired = true;
+        }
+
+        if (anonymous) {
+            // If the security subsystem indicates that the session was established by an anonymous user, Session.SigningRequired MUST be set to FALSE.
+            signingRequired = false;
+        }
+        if (guest && connection.getConfig().isSigningRequired()) {
+            throw new SMBRuntimeException("Cannot require message signing when authenticating with a guest account");
```

Comment:
Could we use a really specific exception for this case that extends `SMBRuntimeException`? It will help with use cases where the client may wish to open a new connection without signing should they discover it's a guest account. i.e. `SMBGuestSigningViolationException` or something?

Reply:
Specifically the logic I'd like to capture is:
```java
Session session = null;
try {
   session = conn.authenticate(auth);
} catch (SMBGuestSigningViolationException e) {
   session.close();
   conn.close();
   conn = // No signing this time, and try again
}
```

The reason for this is I'd like to enforce signing in all cases *except* for guest, where it is not applicable.